File: generate-ssh-keys/generateSSHKey.py
import subprocess
import os
from pathlib import Path
from termcolor import colored
import sys


def ensure_executable(script_path):
    script_file = Path(script_path)
    if not os.access(script_file, os.X_OK):
        script_file.chmod(0o755)


def run_script(script_path, args=[]):
    ensure_executable(script_path)
    result = subprocess.run(
        ["bash", "-c", f"{script_path} {' '.join(args)}"],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        print(colored(f"Error: {result.stderr}", "red"))
        return False
    print(colored(result.stdout, "green"))
    return True


def main():
    ans = input(colored("Set up Git SSH Key? y(es) / n(o): ", "cyan")).strip().lower()
    if ans not in ["y", "yes"]:
        print(colored("No action taken.", "yellow"))
        sys.exit()

    email = input(colored("Enter your email for the SSH key: ", "cyan")).strip()

    script_dir = Path(__file__).parent
    ssh_dir = Path.home() / ".ssh"
    pub_key_path = ssh_dir / "id_ed25519.pub"

    try:
        remove_script_path = script_dir / "remove_ssh.sh"
        setup_script_path = script_dir / "setup_ssh.sh"

        if pub_key_path.exists():
            decision = (
                input(
                    colored(
                        "SSH key exists. Do you want to recreate it? (y(es)/n(o)): ",
                        "cyan",
                    )
                )
                .strip()
                .lower()
            )
            if decision in ["yes", "y"]:
                if run_script(remove_script_path):
                    run_script(setup_script_path, [email])
            else:
                print(colored("Keeping the existing SSH keys.", "yellow"))
        else:
            run_script(setup_script_path, [email])
    except Exception as e:
        print(colored(f"An error occurred: {e}", "red"))


if __name__ == "__main__":
    main()

File: generate-ssh-keys/generate_SSH_key.mjs
import chalk from "chalk";
import { exec } from "child_process";
import fs from "fs/promises";
import path from "path";
import { createInterface } from "readline";

async function ensureExecutable(scriptPath) {
  try {
    await fs.access(scriptPath, fs.constants.X_OK);
  } catch {
    await fs.chmod(scriptPath, 0o755);
  }
}

async function runScript(scriptPath, args = []) {
  await ensureExecutable(scriptPath); // Ensure the script is executable

  return new Promise((resolve, reject) => {
    exec(
      `bash -c "${scriptPath} ${args.join(" ")}"`,
      (error, stdout, stderr) => {
        if (error) {
          console.error(chalk.red("Error:"), chalk.yellow(stderr));
          reject(error);
          return;
        }
        console.log(chalk.green(stdout));
        resolve(stdout);
      }
    );
  });
}

async function main() {
  const rl = createInterface({ input: process.stdin, output: process.stdout });

  rl.question(chalk.cyan("Set up Git SSH Key? y(es) / n(o): "), async (ans) => {
    if (!["y", "yes"].includes(ans.trim().toLowerCase())) {
      console.log(chalk.yellow("No action taken."));
      rl.close();
      return;
    }

    rl.question(
      chalk.cyan("Enter your email for the SSH key: "),
      async (email) => {
        const scriptDir = path.dirname(new URL(import.meta.url).pathname);
        const sshDir = path.join(
          process.env.HOME || process.env.USERPROFILE,
          ".ssh"
        );
        const pubKeyPath = path.join(sshDir, "id_ed25519.pub");

        try {
          const removeScriptPath = path.join(scriptDir, "remove_ssh.sh");
          const setupScriptPath = path.join(scriptDir, "setup_ssh.sh");

          if (
            await fs
              .access(pubKeyPath)
              .then(() => true)
              .catch(() => false)
          ) {
            rl.question(
              chalk.cyan(
                "SSH key exists. Do you want to recreate it? (y(es)/n(o)): "
              ),
              async (decision) => {
                if (["yes", "y"].includes(decision.trim().toLowerCase())) {
                  await runScript(removeScriptPath);
                  await runScript(setupScriptPath, [email]); // Pass email to the script
                } else {
                  console.log(chalk.yellow("Keeping the existing SSH keys."));
                }
                rl.close();
              }
            );
          } else {
            await runScript(setupScriptPath, [email]); // Create SSH key if it doesn't exist
            rl.close();
          }
        } catch (err) {
          console.error(chalk.red("An error occurred:"), err.message);
          rl.close();
        }
      }
    );
  });
}

main();
